### PR TITLE
isprime deterministic to 3317044064679887385961981

### PR DESCRIPTION
**Pull request by oittaa**
#### Brief description of what is fixed or changed

This extends deterministic MR solutions to 3317044064679887385961981 (over 2^81)

#### Other comments
https://arxiv.org/pdf/1509.00864.pdf
https://oeis.org/A014233

#### Release Notes

* ntheory
  * `isprime` extend deterministic solutions to over 2^81

### CODE REVIEW

**Comment by oittaa** [comment]
This might need more benchmarking, but at least on my computer several rounds of Miller-Rabin seems faster than Lucas-test on average. And it might be desirable to have deterministic solutions even though there are no known counterexamples to one round of Miller-Rabin test + Lucas test.

File: sympy/ntheory/primetest.py [hunk 1]
```diff
@@ -621,6 +624,10 @@ def isprime(n):
         return mr(n, [2, 123635709730000, 9233062284813009, 43835965440333360, 761179012939631437, 1263739024124850375])
     if n < 18446744073709551616:
         return mr(n, [2, 325, 9375, 28178, 450775, 9780504, 1795265022])
+    if n < 318665857834031151167461:
```
Comment (smichr) [inline sympy/ntheory/primetest.py:627]: These look suspiciously small relative to the numbers in the other ranges.

**Comment by asmeurer** [comment]
Can you clarify what this does? `isprime(3317044064679887385961981)` returns False already in master.

File: sympy/ntheory/primetest.py (hunk 1 above)
Comment (oittaa) [reply sympy/ntheory/primetest.py:627]: https://arxiv.org/pdf/1509.00864.pdf

**Comment by oittaa** [comment]
> Can you clarify what this does? `isprime(3317044064679887385961981)` returns False already in master.

Yes, there aren't any known pseudoprimes that would pass Baillie–PSW primality test, but to my knowledge it has been proven to be correct "only" up to 2^64. https://arxiv.org/pdf/1509.00864.pdf asserts that Miller-Rabin test with the provided bases is deterministic up to 3317044064679887385961981 (over 2^81). 

The question becomes, is it worth to add complexity to have proven deterministic solutions to around 2^81 instead of 2^64?

**Comment by eagleoflqj** [comment]
> The question becomes, is it worth to add complexity to have proven deterministic solutions to around 2^81 instead of 2^64?

I haven't read the thesis but I think it worth.

**Comment by eagleoflqj** [comment]
I'm not able to judge whether the thesis tells the truth. I just found the table above 4.2 has an obvious typo. `1247050339261*2494100678521` should be `3110269097300703345712981`, not the repeated `3317044064679887385961981`.

**Comment by jksuom** [comment]
> These look suspiciously small relative to the numbers in the other ranges.

It seems that small bases do not make the test less reliable, only less efficient as the range is shorter or the list of bases longer. That can be seen by comparing the tests of [wikipedia](https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test#Testing_against_small_sets_of_bases):

if n < 2,047, it is enough to test a = 2;
if n < 1,373,653, it is enough to test a = 2 and 3;
if n < 9,080,191, it is enough to test a = 31 and 73;
if n < 25,326,001, it is enough to test a = 2, 3, and 5;

with more efficient bases like these: https://github.com/sympy/sympy/blob/68c37df362e8585d72fc7ef490013bb8eff16e3e/sympy/ntheory/primetest.py#L610-L615
So it seems that there only remains this question:
> is it worth to add complexity to have proven deterministic solutions to around 2^81 instead of 2^64?

**Comment by smichr** [comment]
OK, it is a small added complexity to make this deterministic to a significantly larger number.

**Pull request merged**